pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/upsertcompaction/UpsertCompactionTaskGenerator.java

```diff
@@ -0,0 +1,268 @@
+
+      if (completedSegments.isEmpty()) {
+        LOGGER.info("No completed segments were eligible for compaction for table: {}", tableNameWithType);
+        continue;
+      }
+
+      // get server to segment mappings
+      Map<String, List<String>> serverToSegments = _clusterInfoAccessor.getServerToSegmentsMap(tableNameWithType);
+      PinotHelixResourceManager pinotHelixResourceManager = _clusterInfoAccessor.getPinotHelixResourceManager();
+      BiMap<String, String> serverToEndpoints;
+      try {
+        serverToEndpoints = pinotHelixResourceManager.getDataInstanceAdminEndpoints(serverToSegments.keySet());
+      } catch (InvalidConfigException e) {
+        throw new RuntimeException(e);
+      }
+
+      Map<String, SegmentZKMetadata> completedSegmentsMap = completedSegments.stream()
+          .collect(Collectors.toMap(SegmentZKMetadata::getSegmentName, Function.identity()));
+
+      List<String> validDocIdUrls;
+      try {
+        validDocIdUrls = getValidDocIdMetadataUrls(
+            serverToSegments, serverToEndpoints, tableNameWithType, completedSegmentsMap);
+      } catch (URISyntaxException e) {
+        throw new RuntimeException(e);
+      }
+
+      // request the urls from the servers
+      CompletionServiceHelper completionServiceHelper = new CompletionServiceHelper(
+            Executors.newCachedThreadPool(), new MultiThreadedHttpConnectionManager(), serverToEndpoints.inverse());
+      CompletionServiceHelper.CompletionServiceResponse serviceResponse =
+            completionServiceHelper.doMultiGetRequest(validDocIdUrls, tableNameWithType, false, 3000);
+
+      // only compact segments that exceed the threshold
+      double invalidRecordsThresholdPercent =
+          Double.parseDouble(compactionConfigs.getOrDefault(UpsertCompactionTask.INVALID_RECORDS_THRESHOLD_PERCENT,
+              String.valueOf(DEFAULT_INVALID_RECORDS_THRESHOLD_PERCENT)));
+      List<SegmentZKMetadata> segmentsForCompaction = new ArrayList<>();
+      List<String> segmentsForDeletion = new ArrayList<>();
+      for (Map.Entry<String, String> streamResponse : serviceResponse._httpResponses.entrySet()) {
+        JsonNode allValidDocIdMetadata;
+        try {
+          allValidDocIdMetadata = JsonUtils.stringToJsonNode(streamResponse.getValue());
+        } catch (IOException e) {
+          LOGGER.error("Unable to parse validDocIdMetadata response for: {}", streamResponse.getKey());
+          continue;
+        }
+        Iterator<JsonNode> iterator = allValidDocIdMetadata.elements();
+        while (iterator.hasNext()) {
+          JsonNode validDocIdMetadata = iterator.next();
+          double invalidRecordCount = validDocIdMetadata.get("totalInvalidDocs").asDouble();
+          String segmentName = validDocIdMetadata.get("segmentName").asText();
+          SegmentZKMetadata segment = completedSegmentsMap.get(segmentName);
+          double invalidRecordPercent = (invalidRecordCount / segment.getTotalDocs()) * 100;
```

Comment:
Let's change the logic to compute based on the `invalidRecordCount` instead of `invalidRecordPercent`.

We can set the `invalidRecordsThresholdCount` to be 

```
long invalidRecordsThresholdCount;
if (percent config is used) {
  invalidRecordsThresholdCount = segment.getTotalDocs() * config.(PERCENT_CONFIG) // use some rounding
} else {
  invalidRecordsThresholdCount = config.get(COUNT_CONFIG); // if this is null, use default count config (e.g. 10k)
}
```